File: src/main/java/com/baidu/algorithm/bfs/_133_Clone_Graph.py
# Definition for a undirected graph node
class UndirectedGraphNode:
    def __init__(self, x):
        self.label = x
        self.neighbors = []


class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if not node:
            return None

        node_cy = UndirectedGraphNode(node.label)
        mp = {node: node_cy}
        stack = [node]

        while stack:
            cur_node = stack.pop()
            for neigh in cur_node.neighbors:
                if neigh not in mp:
                    node_cy = UndirectedGraphNode(neigh.label)
                    mp[neigh] = node_cy
                    stack.append(neigh)
                mp[cur_node].neighbors.append(mp[neigh])

        return mp[node]

File: src/main/java/com/baidu/algorithm/bfs/_133_Clone_Graph.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.bfs;

import java.util.*;

/**
 * _133_Clone_Graph
 *
 * @author xuhaoran01
 */
public class _133_Clone_Graph {

    class UndirectedGraphNode {
        int label;
        List<UndirectedGraphNode> neighbors;

        UndirectedGraphNode(int x) {
            label = x;
            neighbors = new ArrayList<UndirectedGraphNode>();
        }
    }


    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null) {
            return null;
        }

        Map<UndirectedGraphNode, UndirectedGraphNode> mp = new HashMap<>();
        UndirectedGraphNode nodeDup = new UndirectedGraphNode(node.label);
        mp.put(node, nodeDup);

        Queue<UndirectedGraphNode> queue = new LinkedList<>();
        queue.add(node);

        while (!queue.isEmpty()) {
            UndirectedGraphNode cur = queue.remove();
            UndirectedGraphNode curDup = mp.get(cur);

            for (UndirectedGraphNode neigh : cur.neighbors) {
                if (!mp.containsKey(neigh)) {
                    queue.add(neigh);
                    UndirectedGraphNode neighDup = new UndirectedGraphNode(neigh.label);
                    mp.put(neigh, neighDup);
                }

                curDup.neighbors.add(mp.get(neigh));
            }
        }

        return mp.get(node);
    }
}
